src/com/ms/silverking/cloud/dht/management/OpCompletionTracker.java

```diff
@@ -17,111 +17,111 @@
+    waitTimer = new SimpleTimer(unit, time);
+    for (UUIDBase uuid : uuids) {
+      completionState.putIfAbsent(uuid, new OpCompletionState());
     }
-    
-    public void setComplete(UUIDBase uuid, OpResult result) {
-        OpCompletionState    opCompletionState;
-        
+    try {
+      for (UUIDBase uuid : uuids) {
+        OpCompletionState opCompletionState;
+        boolean complete;
+
         opCompletionState = completionState.get(uuid);
-        if (opCompletionState == null) {
-            // We allow for the slight chance of a leak for now
-            completionState.putIfAbsent(uuid, new OpCompletionState());
-            opCompletionState = completionState.get(uuid);
+        complete = opCompletionState.waitForCompletion(waitTimer.getRemainingMillis(), TimeUnit.MILLISECONDS);
+        if (!complete) {
+          incompleteOps.add(uuid);
         }
-        if (opCompletionState == null) {
-            // This completion state was removed immediately after we added it. Ignore.            
-        } else {
-            opCompletionState.setComplete(result);
+        if (opCompletionState.getResult().hasFailed()) {
+          failedOps.add(uuid);
         }
+      }
+    } finally {
+      for (UUIDBase uuid : uuids) {
+        completionState.remove(uuid);
+      }
     }
-    
-    private static class OpCompletionState {
-        private OpResult    result;
-        private Lock        lock;
-        private Condition    cv;
-        
-        OpCompletionState() {
-            this.lock = new ReentrantLock();
-            this.cv = lock.newCondition();
-            result = OpResult.INCOMPLETE;
-        }
-        
-        boolean waitForCompletion(long time, TimeUnit unit) {
-            Timer    timer;
-            
-            timer = new SimpleTimer(unit, time);
-            lock.lock();
-            try {
-                while (!result.isComplete() && !timer.hasExpired()) {
-                    try {
-                        timer.await(cv);
-                    } catch (InterruptedException e) {
-                    }
-                }
-            } finally {
-                lock.unlock();
-            }
-            return result.isComplete();
-        }
-        
-        void setComplete(OpResult result) {
-            lock.lock();
-            try {
-                if (!this.result.isComplete()) {
-                    this.result = result;
-                }
-                cv.signalAll();
-            } finally {
-                lock.unlock();
-            }
+    return new Pair<>(incompleteOps, failedOps);
+  }
+
+  public void setComplete(UUIDBase uuid, OpResult result) {
+    OpCompletionState opCompletionState;
+
+    opCompletionState = completionState.get(uuid);
+    if (opCompletionState == null) {
+      // We allow for the slight chance of a leak for now
+      completionState.putIfAbsent(uuid, new OpCompletionState());
+      opCompletionState = completionState.get(uuid);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `get()` may not be thread-safe at lines: **69, 70, and 73**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **60**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).